src/main/java/soot/jimple/toolkits/callgraph/VirtualCalls.java

```diff
@@ -106,19 +108,67 @@ public SootMethod resolveNonSpecial(RefType t, NumberedString subSig, boolean ap
       return null;
     }
 
-    SootMethod m = cls.getMethodUnsafe(subSig);
-    if (m != null) {
-      if (!m.isAbstract()) {
-        ret = m;
+    ret = resolveNonSpecialInternal(cls, subSig);
+
+    vtbl.put(subSig, ret);
+    return ret;
+  }
+
+  /**
+   * Resolves a virtual/interface method taking default methods into account.
+   *
+   * @param cls     The class on which the method is called.
+   * @param subSig  Sub signature of the method.
+   * @return        The resolved method. {@code null} if no such method exists.
+   */
+  private SootMethod resolveNonSpecialInternal(SootClass cls, NumberedString subSig) {
+
+    // look for a suitable method in the type directly and in its super types
+    SootClass currentClass = cls;
+    do {
+      SootMethod method = currentClass.getMethodUnsafe(subSig);
+      if (method != null) {
+        if (method.isAbstract()) {
+          return null;
+        } else {
+          return method;
+        }
       }
-    } else {
-      SootClass c = cls.getSuperclassUnsafe();
-      if (c != null) {
-        ret = resolveNonSpecial(c.getType(), subSig);
+    } while ((currentClass = currentClass.getSuperclassUnsafe()) != null);
+
+    // collect interfaces and sort them according to their inheritance distance
+    HashMap<SootClass,Integer> inheritanceDistance = new HashMap<>();
+    currentClass = cls;
+    do {
+      collectInterfaces(currentClass, inheritanceDistance, 0);
+    } while ((currentClass = currentClass.getSuperclassUnsafe()) != null);
+    ArrayList<SootClass> interfaces = new ArrayList<>(inheritanceDistance.keySet());
+    interfaces.sort(Comparator.comparingInt(inheritanceDistance::get));
+
+    // find first default method
+    for (SootClass interf : interfaces) {
+      SootMethod method = interf.getMethodUnsafe(subSig);
+      if (method != null && !method.isAbstract() && !method.isPrivate()) {
+        return method;
       }
     }
-    vtbl.put(subSig, ret);
-    return ret;
+
+    return null;
+  }
+
+  /**
+   * Collects all interfaces which are directly or indirectly implemented by the given class.
+   * Additionally, stores the max inheritance distance between the given class and each interface.
+   *
+   * @param cls                  The class to collect the interfaces for.
+   * @param inheritanceDistance  The map which receives the interfaces together with the max inheritance distance.
+   * @param distance             The current distance from the class of interest.
+   */
+  private void collectInterfaces(SootClass cls, HashMap<SootClass,Integer> inheritanceDistance, int distance) {
+    for (SootClass interf : cls.getInterfaces()) {
+      inheritanceDistance.merge(interf, distance, Integer::max);
```

Comment:
You always store the maximum distance. Shouldn't you store the minimum distance to find the most specific interface?
If a class inherits from the same interface over two different inheritance chains, you would take the farthest away implementation as most specific.

Reply:
The list of all interfaces is sorted in ascending order (line 164). So the "nearest" interface will be at the beginning of the list and will be searched first for an implementation. For one interface, always the maximum distance is stored to make sure it comes in the list after all implemented child interfaces.

Reply:
The correct line number is 146.

Reply:
I see. Thanks for the explanation.